`TestIntegrations/Disconnection/verify_that_lost_connections_to_auth_server_terminate_controlled_connections` flakiness

## Failure

#### Link(s) to logs

- https://console.cloud.google.com/cloud-build/builds/7b377a2f-27d4-4d2b-9d00-c4a891768780?project=ci-account
- v10 https://github.com/gravitational/teleport/actions/runs/4639027594/jobs/8209429583#step:6:1378

#### Relevant snippet

```
    integration_test.go:1325: 
        	Error Trace:	integration_test.go:1325
        	            				integration_test.go:1219
        	Error:      	Async error
        	Test:       	TestIntegrations/Disconnection/Test_6
        	Messages:   	expected EOF, ExitError, or nil, got ssh: could not start shell instead
```

Sadly this has returned: https://github.com/gravitational/teleport/actions/runs/7776666786/job/21204093858?pr=37740

```
=== Failed
=== FAIL: integration TestIntegrations/Disconnection/verify_that_lost_connections_to_auth_server_terminate_controlled_connections (4.37s)
    integration_test.go:2180: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:2184
        	            				/opt/go/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = context canceled
    integration_test.go:2180: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:2185
        	            				/opt/go/src/runtime/asm_amd64.s:1650
        	Error:      	"[]" should have 1 item(s), but has 0
    integration_test.go:2180: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:2180
        	            				/__w/teleport/teleport/integration/integration_test.go:2305
        	            				/__w/teleport/teleport/integration/integration_test.go:2201
        	Error:      	Condition never satisfied
        	Test:       	TestIntegrations/Disconnection/verify_that_lost_connections_to_auth_server_terminate_controlled_connections
        --- FAIL: TestIntegrations/Disconnection/verify_that_lost_connections_to_auth_server_terminate_controlled_connections (4.37s)

=== FAIL: integration TestIntegrations/Disconnection (31.31s)
    --- FAIL: TestIntegrations/Disconnection (31.31s)

=== FAIL: integration TestIntegrations (741.71s)

```

--- a/integration/integration_test.go
+++ b/integration/integration_test.go
@@ -1582,25 +1582,31 @@ func verifySessionJoin(t *testing.T, username string, teleport *helpers.TeleInst
 	// PersonA: SSH into the server, wait one second, then type some commands on stdin:
 	sessionA := make(chan error)
 	openSession := func() {
+		defer close(sessionA)
+
 		cl, err := teleport.NewClient(helpers.ClientConfig{
 			Login:   username,
 			Cluster: helpers.Site,
 			Host:    Host,
+			Stdout:  personA,
+			Stdin:   personA,
 		})
 		if err != nil {
-			sessionA <- trace.Wrap(err)
+			sessionA <- err
 			return
 		}
-		cl.Stdout = personA
-		cl.Stdin = personA
-		// Person A types something into the terminal (including "exit")
-		personA.Type("\aecho hi\n\r\aexit\n\r\a")
-		sessionA <- cl.SSH(context.TODO(), []string{})
+
+		if err := cl.SSH(t.Context(), nil); err != nil {
+			sessionA <- err
+			return
+		}
 	}
 
 	// PersonB: wait for a session to become available, then join:
 	sessionB := make(chan error)
 	joinSession := func() {
+		defer close(sessionB)
+
 		tracker := waitForSessionToBeEstablished(t, site, 1)
 
 		sessionID := tracker.GetSessionID()
@@ -1610,33 +1616,25 @@ func verifySessionJoin(t *testing.T, username string, teleport *helpers.TeleInst
 			Host:    Host,
 		})
 		if err != nil {
-			sessionB <- trace.Wrap(err)
+			sessionB <- err
 			return
 		}
 
-		timeoutCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-		defer cancel()
-
-		ticker := time.NewTicker(100 * time.Millisecond)
-		defer ticker.Stop()
-		for {
-			select {
-			case <-timeoutCtx.Done():
-				sessionB <- timeoutCtx.Err()
-				return
-
-			case <-ticker.C:
-				err := cl.Join(context.TODO(), types.SessionPeerMode, session.ID(sessionID), personB)
-				if err == nil {
-					sessionB <- nil
-					return
-				}
-			}
+		if err := cl.Join(t.Context(), types.SessionPeerMode, session.ID(sessionID), personB); err != nil {
+			sessionB <- err
+			return
 		}
 	}
 
 	go openSession()
 	go joinSession()
+
+	// Wait for both parties to have joined the session and
+	// then enter input and exit the session.
+	waitForSessionToBeEstablished(t, site, 2)
+
+	// Person A types something into the terminal (including "exit")
+	personA.Type("echo hi\n\rexit\n\r")
 
 	// wait for the sessions to end
 	err := waitForError(sessionA, time.Second*10)

